File: packages/typescript/test/behavior.test.ts
/**
 * Normative behavioral suite — spec/behavior-vectors.json (SPECIFICATION.md [N-47]).
 *
 * Every scenario is a published vector, not a hand-written case, so this suite
 * cannot silently drift from the other nine implementations.
 */
import { test } from 'node:test';
import assert from 'node:assert/strict';

import { loadBehaviorVectors, runBehaviorVectors } from './behavior-runner.ts';

const vectors = loadBehaviorVectors();

test('behavior vectors: every applicable scenario passes', async () => {
  const outcome = await runBehaviorVectors(vectors);

  // [N-48]: a runner that silently iterated nothing must not report success.
  assert.equal(
    outcome.executed.length + outcome.skipped.length,
    vectors.counts.scenarios,
    'every published scenario must be either executed or explicitly skipped',
  );
  assert.ok(outcome.executed.length >= vectors.counts.unconditional,
    'every unconditional scenario must be executed');

  // This runtime's strings are UTF-16, so nothing should be skipped here.
  assert.deepEqual(outcome.skipped, [], 'no scenario is inapplicable to Node.js');
});

test('behavior vectors: each scenario runs in isolation and is individually named', async () => {
  for (const scenario of vectors.scenarios) {
    await test(scenario.id, async () => {
      const single = { ...vectors, scenarios: [scenario], counts: { scenarios: 1, unconditional: 1 } };
      const outcome = await runBehaviorVectors(single);
      assert.equal(outcome.executed.length + outcome.skipped.length, 1);
    });
  }
});

File: packages/dart/test/behavior_test.dart
/// Normative behavioral suite — spec/behavior-vectors.json
/// (SPECIFICATION.md [N-47], [N-49]).
///
/// Every scenario is a published vector, not a hand-written case, so this suite
/// cannot silently drift from the other nine implementations.
library;

import 'package:nebula_token/nebula_token.dart' as nt;
import 'package:test/test.dart';

import 'behavior_runner.dart';

void main() {
  final BehaviorVectors vectors = loadBehaviorVectors();

  test('the package implements the spec version the vectors were cut from', () {
    expect(nt.specVersion, vectors.specVersion);
  });

  test('behavior vectors: every applicable scenario passes', () async {
    final RunOutcome outcome = await runBehaviorVectors(vectors);

    // [N-48]: a runner that silently iterated nothing must not report success.
    expect(
      outcome.executed.length + outcome.skipped.length,
      vectors.scenarioCount,
      reason: 'every published scenario must be executed or explicitly skipped',
    );
    expect(
      outcome.executed.length,
      greaterThanOrEqualTo(vectors.unconditionalCount),
      reason: 'every unconditional scenario must be executed',
    );

    // Dart strings are UTF-16, so the one conditional scenario applies here too
    // and nothing is inapplicable. Skips are reported by id when they happen.
    expect(
      outcome.skipped,
      isEmpty,
      reason: 'skipped scenarios: ${outcome.skipped}',
    );
  });

  group('behavior vectors: each scenario in isolation', () {
    for (final Scenario scenario in vectors.scenarios) {
      final String? condition = scenario.condition;
      test(
        '${scenario.id} — ${scenario.title}',
        () async {
          final RunOutcome outcome = await runBehaviorVectors(
            vectors.withOnly(scenario),
          );
          expect(outcome.executed, <String>[scenario.id]);
        },
        skip: condition != null && !satisfiedConditions.contains(condition)
            ? 'runtime does not satisfy "$condition"'
            : null,
      );
    }
  });
}
